test_util/src/lsp.rs

```diff
@@ -166,13 +209,42 @@ impl LspClient {
   where
     R: de::DeserializeOwned,
   {
+    if !self.msg_queue.is_empty() {
+      let mut msg_queue = VecDeque::new();
+      loop {
+        match self.msg_queue.pop_front() {
+          Some(LspMessage::Request(id, method, maybe_params)) => {
+            if let Some(p) = maybe_params {
+              let params = serde_json::from_value(p)?;
+              return Ok((id, method, Some(params)));
+            } else {
+              return Ok((id, method, None));
+            }
+          }
+          Some(msg) => {
+            msg_queue.push_back(msg);
+          }
+          _ => break,
+        }
+      }
+      self.msg_queue = msg_queue;
+    }
```

Comment:
This bit of code is repeated several times... can that be cleaned up somehow?

Reply:
I've tried to clean it up as much as possible, while the blocks look really similar, they are subtly different, but I did my best.